x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java

```diff
@@ -0,0 +1,281 @@
+                        + "] is below the configured minimum number of ["
+                        + decider.getMinNumNodes()
+                        + "] or is zero");
+            }
+            Set<String> timedUpJobs = new HashSet<>();
+            for (String jobId : waitingAnomalyJobs) {
+                long time = anomalyJobsTimeInQueue.compute(jobId, (k, v) -> v == null ? 0L : v + timeSinceLastCheckMs);
+                if (time >= decider.getAnomalyJobTimeInQueue().getMillis()) {
+                    timedUpJobs.add(jobId);
+                }
+            }
+            for (String jobId : waitingAnalysisJobs) {
+                long time = analyticsJobsTimeInQueue.compute(jobId, (k, v) -> v == null ? 0L : v + timeSinceLastCheckMs);
+                if (time >= decider.getAnalysisJobTimeInQueue().getMillis()) {
+                    timedUpJobs.add(jobId);
+                }
+            }
+            if (timedUpJobs.isEmpty() == false) {
+                return new AutoscalingDecision(name(),
+                    AutoscalingDecisionType.SCALE_UP,
+                    "jobs " + timedUpJobs + " have been waiting for assignment too long");
+            }
+        }
+        return new AutoscalingDecision(name(), AutoscalingDecisionType.NO_SCALE, "no jobs have waited long enough for assignment");
+    }
+
+    AutoscalingDecision checkForScaleDown(MlAutoscalingDeciderConfiguration decider,
+                                          List<DiscoveryNode> nodes,
+                                          ClusterState clusterState) {
+        if (nodes.size() == decider.getMinNumNodes()) {
+            return new AutoscalingDecision(name(),
+                AutoscalingDecisionType.NO_SCALE,
+                "| already at configured minimum node count ["
+                    + decider.getMinNumNodes()
+                    +"]");
+        }
+        // kick of a refresh if our data around memory usage is stale.
+        if (nodeLoadDetector.getMlMemoryTracker().isRecentlyRefreshed() == false) {
+            logger.debug("job memory tracker is stale. Request refresh before making scaling decision.");
+            nodeLoadDetector.getMlMemoryTracker().asyncRefresh();
+            return new AutoscalingDecision(
+                name(),
+                AutoscalingDecisionType.NO_SCALE,
+                "job memory tracker is stale. Unable to make accurate scale down recommendation");
+        }
+        // TODO: remove in 8.0.0
+        boolean allNodesHaveDynamicMaxWorkers = clusterState.getNodes().getMinNodeVersion().onOrAfter(Version.V_7_2_0);
+        List<NodeLoadDetector.NodeLoad> nodeLoads = new ArrayList<>();
+        boolean isMemoryAccurateFlag = true;
+        for (DiscoveryNode node : nodes) {
+            NodeLoadDetector.NodeLoad nodeLoad = nodeLoadDetector.detectNodeLoad(clusterState,
+                allNodesHaveDynamicMaxWorkers,
+                node,
+                maxOpenJobs,
+                maxMachineMemoryPercent,
```

Comment:
One nasty thing is that in Cloud the value of `max_machine_memory_percent` will change as the nodes are scaled: https://github.com/elastic/cloud/blob/65712c90f8b1bc67df949c7b711d5aff5e1f43ea/scala-services/constructor/src/main/scala/no/found/constructor/steps/ml/OptimizeMlClusterSettings.scala#L191-L194

As the ML nodes get bigger the JVM occupies a smaller proportion of the node, so a bigger proportion is available for ML native processes.  So as the nodes get bigger `max_machine_memory_percent` increases.  This is potentially suboptimal but not a disaster when scaling up - we might scale up one step more than necessary.  But after scaling down it might mean the jobs don't fit in the permitted memory. 😱 

Ideally we would do the calculation with the value of `max_machine_memory_percent` that would apply at the size we were scaling _to_, both when scaling up and down.  But this requires that we know how `max_machine_memory_percent` changes as the node size changes.  So, either we have to duplicate the Cloud formula in ES code (and assume only Cloud will use autoscaling), or add a setting that is an alternative way to specify `max_machine_memory` as an arbitrary function of node size and JVM heap size instead of as a percentage of node size.  That would actually make the Cloud code simpler in the future, as they could then just put their formula in a static setting instead of having to dynamically change the value for different node sizes.

Reply:
> or add a setting that is an alternative way to specify `max_machine_memory` as an arbitrary function of node size and JVM heap size instead of as a percentage of node size.

Yeesh...this sounds error prone :(. 

I also don't think this is possible given the current ways we parse settings. How would users even provide the function for use to use? Just a string setting that we parse into a script D: ???

This is a pickle...